src/Raven.Server/Documents/TimeSeries/TimeSeriesPolicyRunner.cs

```diff
@@ -0,0 +1,683 @@
+        {
+            var skip = 0;
+            while (true)
+            {
+                Cts.Token.ThrowIfCancellationRequested();
+
+                var cmd = new AddedNewRollupPoliciesCommand(collectionName, prev, policy, skip);
+                await _database.TxMerger.Enqueue(cmd);
+
+                if (cmd.Marked < AddedNewRollupPoliciesCommand.BatchSize)
+                    break;
+
+                skip += cmd.Marked;
+            }
+        }
+
+        internal async Task RunRollUps()
+        {
+            var now = DateTime.UtcNow;
+            try
+            {
+                var states = new List<RollUpState>();
+                using (_database.DocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                {
+                    while (true)
+                    {
+                        context.Reset();
+                        context.Renew();
+
+                        Stopwatch duration;
+                        using (context.OpenReadTransaction())
+                        {
+                            PrepareRollUps(context, now, 1024, states, out duration);
+                            if (states.Count == 0)
+                                return;
+                        }
+
+                        Cts.Token.ThrowIfCancellationRequested();
+
+                        var topology = _database.ServerStore.LoadDatabaseTopology(_database.Name);
+                        var isFirstInTopology = string.Equals(topology.Members.FirstOrDefault(), _database.ServerStore.NodeTag, StringComparison.OrdinalIgnoreCase);
+
+                        var command = new RollupTimeSeriesCommand(Configuration, states, isFirstInTopology);
+                        await _database.TxMerger.Enqueue(command);
+                        if (command.RolledUp == 0)
+                            break;
+
+                        states.Clear();
+
+                        if (Logger.IsInfoEnabled)
+                            Logger.Info($"Successfully aggregated {command.RolledUp:#,#;;0} time-series within {duration.ElapsedMilliseconds:#,#;;0} ms.");
+                    }
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // this will stop processing
+                throw;
+            }
+            catch (Exception e)
+            {
+                if (Logger.IsOperationsEnabled)
+                        Logger.Operations($"Failed to roll-up time series for '{_database.Name}' which are older than {now}", e);
+            }
+        }
+
+        internal async Task DoRetention()
+        {
+            var topology = _database.ServerStore.LoadDatabaseTopology(_database.Name);
+            var isFirstInTopology = string.Equals(topology.Members.FirstOrDefault(), _database.ServerStore.NodeTag, StringComparison.OrdinalIgnoreCase);
+            if (isFirstInTopology == false)
```

Comment:
Do we want to go with first? Isn't that the most busiest one? It probably it the most up to date, but do we care about that?

Reply:
But there is a point that we are putting more and more load on the primary.. for now I leave it as is.